packages/waku/src/lib/renderers/html.ts

```diff
@@ -0,0 +1,269 @@
+    const f = (s) => c.enqueue(typeof s === 'string' ? t.encode(s) : s);
+    d.forEach(f);
+    d.push = f;
+    if (document.readyState === 'loading') {
+      document.addEventListener('DOMContentLoaded', () => c.close());
+    } else {
+      c.close();
+    }
+  }
+})))
+`
+  .split('\n')
+  .map((line) => line.trim())
+  .join('');
+
+const CLOSING_HEAD = '</head>';
+const CLOSING_BODY = '</body>';
+
+const injectHtmlHead = (
+  urlForFakeFetch: string,
+  htmlHead: string,
+  mainJsPath: string, // for DEV only, pass `''` for PRD
+) => {
+  const modifyHeadAndBody = (data: string) => {
+    const closingHeadIndex = data.indexOf(CLOSING_HEAD);
+    let [head, body] =
+      closingHeadIndex === -1
+        ? ['<head>' + CLOSING_HEAD, data]
+        : [
+            data.slice(0, closingHeadIndex + CLOSING_HEAD.length),
+            data.slice(closingHeadIndex + CLOSING_HEAD.length),
+          ];
+    head =
+      head.slice(0, -CLOSING_HEAD.length) +
+      DEFAULT_HTML_HEAD +
+      htmlHead +
+      CLOSING_HEAD;
+    const matchPrefetched = head.match(
+      // HACK This is very brittle
+      /(.*<script[^>]*>\nglobalThis\.__WAKU_PREFETCHED__ = {\n)(.*?)(\n};.*)/s,
+    );
+    if (matchPrefetched) {
+      head =
+        matchPrefetched[1] +
+        `  '${urlForFakeFetch}': ${fakeFetchCode},` +
+        matchPrefetched[3];
+    }
+    let code = `
+globalThis.__WAKU_HYDRATE__ = true;
+`;
+    if (!matchPrefetched) {
+      code += `
+globalThis.__WAKU_PREFETCHED__ = {
+  '${urlForFakeFetch}': ${fakeFetchCode},
+};
+`;
+    }
+    if (code) {
+      head =
+        head.slice(0, -CLOSING_HEAD.length) +
+        `<script type="module" async>${code}</script>` +
+        CLOSING_HEAD;
+    }
+    if (mainJsPath) {
+      const closingBodyIndex = body.indexOf(CLOSING_BODY);
+      const [firstPart, secondPart] =
+        closingBodyIndex === -1
+          ? [body, '']
+          : [body.slice(0, closingBodyIndex), body.slice(closingBodyIndex)];
+      body =
+        firstPart +
+        `<script src="${mainJsPath}" async type="module"></script>` +
+        secondPart;
+    }
+    return head + body;
+  };
+  const encoder = new TextEncoder();
+  const decoder = new TextDecoder();
+  let headSent = false;
+  let data = '';
+  return new TransformStream({
+    transform(chunk, controller) {
+      if (!(chunk instanceof Uint8Array)) {
+        throw new Error('Unknown chunk type');
+      }
+      data += decoder.decode(chunk);
+      if (!headSent) {
+        if (!/<body[^>]*>/.test(data)) {
+          return;
+        }
+        headSent = true;
+        data = modifyHeadAndBody(data);
+      }
+      controller.enqueue(encoder.encode(data));
+      data = '';
+    },
+    flush(controller) {
+      if (!headSent) {
+        headSent = true;
+        data = modifyHeadAndBody(data);
+        controller.enqueue(encoder.encode(data));
+        data = '';
+      }
+    },
+  });
+};
+
+// HACK for now, do we want to use HTML parser?
```

Comment:
html parser sounds nice for this

Reply:
yeah, I thought so and looked for something, but performance may go down with the parser.